clients/src/main/java/org/apache/kafka/common/telemetry/internals/ClientTelemetryProvider.java

```diff
@@ -0,0 +1,149 @@
+import org.apache.kafka.common.Configurable;
+import org.apache.kafka.common.metrics.MetricsContext;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class ClientTelemetryProvider implements Configurable {
+
+    public static final String DOMAIN = "org.apache.kafka";
+    // Client metrics tags
+    public static final String CLIENT_RACK = "client_rack";
+    public static final String GROUP_ID = "group_id";
+    public static final String GROUP_INSTANCE_ID = "group_instance_id";
+    public static final String GROUP_MEMBER_ID = "group_member_id";
+    public static final String TRANSACTIONAL_ID = "transactional_id";
+
+    private static final String PRODUCER_NAMESPACE = "kafka.producer";
+    private static final String CONSUMER_NAMESPACE = "kafka.consumer";
+
+    private static final Map<String, String> PRODUCER_CONFIG_MAPPING = new HashMap<>();
+    private static final Map<String, String> CONSUMER_CONFIG_MAPPING = new HashMap<>();
+
+    private volatile Resource resource = null;
+    private Map<String, ?> config = null;
+
+    // Mapping of config keys to telemetry keys. Contains only keys which can be fetched from config.
+    // Config like group_member_id is not present here as it is not fetched from config.
+    static {
+        PRODUCER_CONFIG_MAPPING.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, ClientTelemetryProvider.TRANSACTIONAL_ID);
+
+        CONSUMER_CONFIG_MAPPING.put(ConsumerConfig.GROUP_ID_CONFIG, ClientTelemetryProvider.GROUP_ID);
+        CONSUMER_CONFIG_MAPPING.put(ConsumerConfig.GROUP_INSTANCE_ID_CONFIG, ClientTelemetryProvider.GROUP_INSTANCE_ID);
+    }
+
+    @Override
+    public synchronized void configure(Map<String, ?> configs) {
+        this.config = configs;
+    }
+
+    /**
+     * Validate that all the data required for generating correct metrics is present.
+     *
+     * @param metricsContext {@link MetricsContext}
+     * @return false if all the data required for generating correct metrics is missing, true
+     * otherwise.
+     */
+    boolean validate(MetricsContext metricsContext) {
+        return ClientTelemetryUtils.validateRequiredResourceLabels(metricsContext.contextLabels());
+    }
+
+    /**
+     * Sets the metrics tags for the service or library exposing metrics. This will be called before
+     * {@link org.apache.kafka.common.metrics.MetricsReporter#init(List)} and may be called anytime
+     * after that.
+     *
+     * @param metricsContext {@link MetricsContext}
+     */
+    synchronized void contextChange(MetricsContext metricsContext) {
+        final Resource.Builder resourceBuilder = Resource.newBuilder();
+
+        final String namespace = metricsContext.contextLabels().get(MetricsContext.NAMESPACE);
+        if (PRODUCER_NAMESPACE.equals(namespace)) {
+            // Add producer resource labels.
+            PRODUCER_CONFIG_MAPPING.forEach((configKey, telemetryKey) -> {
+                if (config.containsKey(configKey)) {
```

Comment:
do we need to log debug if the configKey does not exist? This might be useful to debug purpose.